src/main/java/at/ac/tuwien/kr/alpha/api/query/AnswerSetQuery.java

```diff
@@ -0,0 +1,189 @@
+package at.ac.tuwien.kr.alpha.api.query;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import at.ac.tuwien.kr.alpha.common.AnswerSet;
+import at.ac.tuwien.kr.alpha.common.Predicate;
+import at.ac.tuwien.kr.alpha.common.atoms.Atom;
+import at.ac.tuwien.kr.alpha.common.terms.ConstantTerm;
+import at.ac.tuwien.kr.alpha.common.terms.FunctionTerm;
+import at.ac.tuwien.kr.alpha.common.terms.Term;
+
+/**
+ * A query for ASP atoms matching a set of filter predicates.
+ */
+public final class AnswerSetQuery implements java.util.function.Predicate<Atom> {
+
+	private final Predicate predicate;
+	private Map<Integer, java.util.function.Predicate<Term>> filters = new HashMap<>();
+
+	private AnswerSetQuery(Predicate pred) {
+		this.predicate = pred;
+	}
+
+	/**
+	 * Creates a new AnswerSetQuery that will match all atoms that are instances of the given {@link Predicate}.
+	 * 
+	 * @param predicate the predicate to match against
+	 * @return a new AnswerSetQuery matching against the given predicate
+	 */
+	public static AnswerSetQuery forPredicate(Predicate predicate) {
+		return new AnswerSetQuery(predicate);
+	}
+
+	/**
+	 * Adds a new filter to this AnswerSetQuery.
+	 * For an atom <code>a(t1, ..., tn)</code>, the term at index <code>termIdx</code> will be tested against the given filter predicate.
+	 * 
+	 * @param termIdx the index of the term to test
+	 * @param filter  the test predicate to use on terms
+	 * @return this AnswerSetQuery with the additional filter added
+	 */
+	public AnswerSetQuery withFilter(int termIdx, java.util.function.Predicate<Term> filter) {
+		if (termIdx >= this.predicate.getArity()) {
+			throw new IndexOutOfBoundsException(
+					"Predicate " + this.predicate.getName() + " has arity " + this.predicate.getArity() + ", term index " + termIdx + " is invalid!");
+		}
+		if (this.filters.containsKey(termIdx)) {
+			java.util.function.Predicate<Term> currFilter = this.filters.get(termIdx);
+			this.filters.put(termIdx, currFilter.and(filter));
+		} else {
+			this.filters.put(termIdx, filter);
+		}
+		return this;
+	}
+
+	/**
+	 * Convenience method - adds a filter to match names of symbolic constants against a string.
+	 * 
+	 * @param termIdx
+	 * @param str
+	 * @return
+	 */
+	public AnswerSetQuery withConstantEquals(int termIdx, String str) {
+		return this.withFilter(termIdx, AnswerSetQuery.constantTermEquals(str));
```

Comment:
The method `constantTermEquals` is only used here and can be inlined as follows:
```suggestion
		return this.withFilter(termIdx, (t) -> AnswerSetQuery.constantTermEquals(t, str));
```